packages/vscode-extension/src/builders/PlatformBuildCache.ts

```diff
@@ -0,0 +1,152 @@
+import path from "path";
+import fs from "fs";
+import { createFingerprintAsync } from "@expo/fingerprint";
+import { Logger } from "../Logger";
+import { extensionContext, getAppRootFolder } from "../utilities/extensionContext";
+import { DevicePlatform } from "../common/DeviceManager";
+import { IOSBuildResult } from "./buildIOS";
+import { AndroidBuildResult } from "./buildAndroid";
+import { getLaunchConfiguration } from "../utilities/launchConfiguration";
+import { runFingerprintScript } from "./customBuild";
+import { CancelToken } from "./cancelToken";
+import { calculateMD5 } from "../utilities/common";
+import { BuildResult } from "./BuildManager";
+
+const ANDROID_BUILD_CACHE_KEY = "android_build_cache";
+const IOS_BUILD_CACHE_KEY = "ios_build_cache";
+
+const IGNORE_PATHS = [
+  path.join("android", ".gradle/**/*"),
+  path.join("android", "build/**/*"),
+  path.join("android", "app", "build/**/*"),
+  path.join("ios", "build/**/*"),
+  "**/node_modules/**/android/.cxx/**/*",
+  "**/node_modules/**/.gradle/**/*",
+  "**/node_modules/**/android/build/intermediates/cxx/**/*",
+];
+
+export type BuildCacheInfo = {
+  fingerprint: string;
+  buildHash: string;
+  buildResult: AndroidBuildResult | IOSBuildResult;
+};
+
+export class PlatformBuildCache {
+  private cacheKey: string;
+
+  constructor(
+    private readonly platform: DevicePlatform,
+    private readonly cancelToken: CancelToken
```

Comment:
cancel tokens are transient and shouldn't be kept like that. Their purpose is to be able to pass them to async methods, such that they can be intercepted. Storing cancelToken like that defeats this purpose, because who is now in the control of cancelling it? 

We should likely pass cancelTokens to methods that needs them directly, such that the caller is responsible for cancelling. 
Alternatively we can create new cancel tokens for the calls that require them and expose some cancel method from this class which may be called by the caller.